feat(ai-gateway): AI Lakera Guard overview (#5788)
Kong/developer.konghq.com: app/_ai_gateway_policies/ai-lakera-guard/index.md

  - konnect
products:
  - ai-gateway
content_type: policy
---

> The AI Lakera Guard Policy evaluates requests and responses that pass through {{site.ai_gateway}} to Large Language Models (LLMs). It uses the Lakera Guard SaaS service to detect safety policy violations and block unsafe content before it reaches upstream LLMs or returns to clients. The AI Lakera Guard Policy supports multiple inspection modes and guards both inbound prompts and outbound model outputs.

Reviewer: ```suggestion
The AI Lakera Guard Policy evaluates requests and responses that pass through {{site.ai_gateway}} to Large Language Models (LLMs). It uses the [Lakera Guard SaaS service](https://www.lakera.ai/) to detect safety policy violations and block unsafe content before it reaches upstream LLMs or returns to clients. The AI Lakera Guard Policy supports multiple inspection modes and guards both inbound prompts and outbound model outputs.
```

* **Request phase**: Inspection occurs **before** any data leaves the gateway toward the target LLM. The AI Lakera Guard Policy buffers the full request body in memory, extracts the fields that the AI Lakera Guard Policy can evaluate, and sends them for inspection.
* **Response phase (buffered)**: Inspection occurs **before** any byte is transmitted back toward the client. The AI Lakera Guard Policy buffers the full upstream response in memory, extracts the response fields that Lakera Guard can evaluate, and inspects them. This occurs before {{site.ai_gateway}} sends any part of the response back to the client.
* **Response phase (per-frame)**: The AI Lakera Guard Policy runs during streaming responses like Server-Sent Events. {{site.ai_gateway}} processes the response in chunks, buffering each frame in memory as it arrives. When enough data is available to extract an evaluable segment, the AI Lakera Guard Policy inspects that segment with Lakera Guard before forwarding the frame to the client.

> The AI Lakera Guard Policy inspects request and response bodies for routes that use supported model interaction formats. It skips inspection on response types that are not text responses based on Lakera Guard's current product limitations.

Reviewer: I found the second sentence to be quite ambiguous so this is my suggestion:

```suggestion
The AI Lakera Guard Policy inspects request and response bodies for routes that use supported model interaction formats. It skips inspection on non-text response types, which Lakera Guard does not currently support.
```

{% endtable %}

## Logging

You can use the [logging capabilities](/ai-gateway/ai-audit-log-reference/) of the AI Lakera Guard Policy to monitor the inspection process and understand the detected violations. For the full list of log fields, see the [{{site.ai_gateway}} audit log reference](/ai-gateway/ai-audit-log-reference/#ai-lakera-guard-logs).

The AI Lakera Guard Policy provides detailed logging and controls over how violations are reported:
* **SaaS platform logging**: All inspected requests, responses, and chats are made available on the Lakera SaaS platform.
* **{{site.ai_gateway}} logging**: {{site.ai_gateway}} logs all request and response **Lakera request UUIDs** to the standard logging subsystem.
* **Unsupported logging outputs**: [Prometheus](/ai-gateway/policies/prometheus/), or [OpenTelemetry](/ai-gateway/policies/opentelemetry/).
* **Logging outputs**: [HTTP Log](/ai-gateway/policies/http-log/), [File Log](/ai-gateway/policies/file-log/), and [TCP Log](/ai-gateway/policies/tcp-log/).

By default, the AI Lakera Guard Policy doesn't tell clients why their request was blocked. However, this information is always logged to {{site.ai_gateway}} logs for administrators.

> To change this behavior, use `reveal_failure_categories: true`. If activated, you'll receive a JSON response including a breakdown array that details the specific `detector_type` that caused the failure.

Reviewer: ```suggestion
To change this behavior, use `reveal_failure_categories: true`. If activated, the client receives a JSON response including a breakdown array that details the specific `detector_type` that caused the failure.
```

To log the raw content of blocked requests and responses, enable [`config.log_blocked_content`](/ai-gateway/policies/ai-lakera-guard/reference/#schema--config-log-blocked-content). When enabled, the blocked prompt or response body appears under `ai.proxy.lakera-guard.input_faulty_prompt` and `ai.proxy.lakera-guard.output_faulty_response` in the log entry.

### Standard logging subsystem example

When a request passes all guardrails, the log includes processing latency and the request UUID:

```json
"ai": {
  "proxy": {
    "lakera-guard": {
      "input_processing_latency": 72,
      "lakera_service_url": "https://api.lakera.ai/v2/guard",
      "input_request_uuid": "a1b2c3d4-5e6f-7a8b-9c0d-1e2f3a4b5c6d",
      "lakera_project_id": "project-1234567890"
    }
  }
}
```

### Violations log example

> When a request is blocked, the log captures the violation reason, detector details, and the blocking AI Policy name, AI Consumer ID, and a running trigger counter:

Reviewer: ```suggestion
When the guardrails block a request, the log captures the violation reason, detector details, and the blocking AI Policy name, AI Consumer ID, and a running trigger counter:
```
